### PR TITLE
bug fixes & !ar search-by-rank

### PR DESCRIPTION
**Pull request by dreamingkills**
coca cola 마시다

### CODE REVIEW

File: src/commands/Lastfm/Friends/Commands/Scrobbles.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class Scrobbles extends FriendsChildCommand<typeof args> {
   idSeed = "nature sohee";
 
   description = "Shows how many scrobbles your friends have";
-  aliases = ["s"];
+  aliases = ["s", "scromples"];
```
Comment (jivison) [inline src/commands/Lastfm/Friends/Commands/Scrobbles.ts:22]: This is the `!friends scrobbles` command 😅

File: src/commands/Lastfm/Rank/AlbumAt.ts [hunk 2]
```diff
@@ -31,30 +28,8 @@ export default class AlbumAt extends LastFMBaseCommand<typeof args> {
     rank: new validators.NumberValidator({ whole: true }),
   };
 
-  async run() {
-    const rank = this.parsedArguments.rank;
+  // merged with !albumrank
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumAt.ts:31]:
```suggestion
```

File: src/commands/Lastfm/Rank/AlbumRank.ts [hunk 3]
```diff
@@ -41,47 +50,66 @@ export default class AlbumRank extends LastFMBaseCommand<typeof args> {
...
-      limit: 1000,
+      limit: topArguments.limit,
+      page: topArguments.page,
     });
 
-    const rank = topAlbums.albums.findIndex(
+    const originalAlbumCount = albums.length;
+
+    if (albums.length === 0)
+      throw new LogicError("You haven't scrobbled that many albums!");
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumRank.ts:66]: This should look like:
```ts
if (!albums.length) {
    throw new RankTooHighError("album");
}
```

or something, maybe you can think of a better name for the error. The error should go in the `errors.ts` file

File: src/commands/Lastfm/Rank/AlbumRank.ts [hunk 4]
```diff
@@ -41,47 +50,66 @@ export default class AlbumRank extends LastFMBaseCommand<typeof args> {
...
+
+    albums = this.getSlice({
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumRank.ts:68]: This should be its own variable called `slicedAlbums`

File: src/commands/Lastfm/Rank/AlbumRank.ts [hunk 5]
```diff
@@ -41,47 +50,66 @@ export default class AlbumRank extends LastFMBaseCommand<typeof args> {
...
+        a.artist.name.toLowerCase() === artist.toLowerCase()
     );
 
-    if (rank === -1) {
+    if ((rank || index) === -1) {
       throw new LogicError(
         `That album wasn't found in ${
           perspective.possessive
-        } top ${displayNumber(topAlbums.albums.length, "album")}`
+        } top ${displayNumber(originalAlbumCount, "album")}`
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumRank.ts:85]: With the above change you can just call `albums.length`

File: src/commands/Lastfm/Rank/AlbumRank.ts [hunk 6]
```diff
@@ -41,47 +50,66 @@ export default class AlbumRank extends LastFMBaseCommand<typeof args> {
...
       .setTitle(
-        `Albums around ${topAlbums.albums[rank].name} in ${perspective.possessive} library`
+        `Albums around ${
+          shouldSearchByRank
+            ? `rank ${displayNumber(rank)}`
+            : albums[index].name
+        } in ${perspective.possessive} library`
       )
-      .setURL(LastfmLinks.libraryAroundAlbum(username, rank))
+      .setURL(LastfmLinks.libraryAroundAlbum(username, albums[0].rank))
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumRank.ts:98]: Is this the right URL? This looks like the n-5th album

File: src/commands/Lastfm/Rank/AlbumRank.ts [hunk 7]
```diff
@@ -41,47 +50,66 @@ export default class AlbumRank extends LastFMBaseCommand<typeof args> {
...
       .setDescription(
         displayNumberedList(
-          topAlbums.albums.slice(start, stop).map((val, idx) => {
+          albums.map((val) => {
             const display = `${italic(val.name)} by ${sanitizeForDiscord(
-              val.artist.name
+              val instanceof TopArtist ? val.name : val.artist.name
```
Comment (jivison) [inline src/commands/Lastfm/Rank/AlbumRank.ts:103]: In what circumstance would `val` be a `TopArtist`?

File: src/commands/Lastfm/Rank/RankCommand.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+      const index = entities.indexOf(targetEntity!);
+      entities = entities.slice(Math.max(index - 5, 0), index + 6);
+
+      return entities as T;
+    }
+
+    if (headWaste !== undefined) entities = entities.slice(headWaste);
+    if (tailWaste !== undefined)
+      entities = entities.slice(0, rank && rank <= 5 ? 11 - tailWaste : 11);
```
Comment (jivison) [inline src/commands/Lastfm/Rank/RankCommand.ts:70]:
```suggestion
    if (headWaste !== undefined) {
      entities = entities.slice(headWaste);
    }
    
    if (tailWaste !== undefined) {
      entities = entities.slice(0, rank && rank <= 5 ? 11 - tailWaste : 11);
    }
```

File: src/commands/Lastfm/Rank/RankCommand.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    } else if (index) {
+      targetEntity = entities[index];
+    } else throw new LogicError("No rank or index specified!");
+
+    const centerIndex = entities.indexOf(targetEntity);
+
+    start = Math.max(0, centerIndex - 5);
+    stop = Math.min(centerIndex + 6, entities.length);
+
+    return entities.slice(start, stop) as T;*/
```
Comment (jivison) [inline src/commands/Lastfm/Rank/RankCommand.ts:88]:
```suggestion
```

File: src/commands/Lastfm/Rank/TrackAt.ts [hunk 10]
```diff
@@ -1,11 +1,8 @@
...
-import { bold, italic } from "../../../helpers/discord";
 import { NumberArgument } from "../../../lib/context/arguments/argumentTypes/NumberArgument";
 import { standardMentions } from "../../../lib/context/arguments/mentionTypes/mentions";
 import { ArgumentsMap } from "../../../lib/context/arguments/types";
 import { Validation } from "../../../lib/validation/ValidationChecker";
 import { validators } from "../../../lib/validation/validators";
-import { displayNumber } from "../../../lib/views/displays";
 import { LastFMBaseCommand } from "../LastFMBaseCommand";
 
 const args = {
```
Comment (jivison) [inline src/commands/Lastfm/Rank/TrackAt.ts:8]: You can delete the `args` for all the archived commands

File: src/commands/Lastfm/Rank/TrackAt.ts [hunk 11]
```diff
@@ -32,30 +29,7 @@ export default class TrackAt extends LastFMBaseCommand<typeof args> {
     rank: new validators.NumberValidator({ whole: true }),
```
Comment (jivison) [inline src/commands/Lastfm/Rank/TrackAt.ts:29]: The validation can go too

File: src/commands/Lastfm/Rank/AlbumRank.ts (hunk 7 above)
Comment (jivison) [reply src/commands/Lastfm/Rank/AlbumRank.ts:103]: ^^ all the above changes should be implemented in `TrackRank` and `ArtistRank`

**Review by jivison — CHANGES_REQUESTED** [review]
Looks pretty good! It would also be great if you could rename everything "rank" to "around". 

Ie. `src/commands/Lastfm/Rank/AlbumRank.ts` -> `src/commands/Lastfm/Around/AlbumAround.ts`. The class names should be changed, but keep in mind the class names are used as aliases, so the "albumaround" alias should be replaced with "albumrank" when the class is renamed. (I hope that makes sense)

File: src/commands/Lastfm/Rank/AlbumAt.ts (hunk 2 above)
Comment (dreamingkills) [reply src/commands/Lastfm/Rank/AlbumAt.ts:31]: oooops im blind  fixed in 37a6997231672cd7dfdd87b18efec760947ac3f2

File: src/commands/Lastfm/Friends/Commands/Scrobbles.ts (hunk 1 above)
Comment (dreamingkills) [reply src/commands/Lastfm/Friends/Commands/Scrobbles.ts]: ooops im blind fixed in 37a6997231672cd7dfdd87b18efec760947ac3f2

**Comment by jivison** [comment]
Looks good!

**Review by jivison — APPROVED** [review]

**Pull request merged**